ntp-udp/src/lib.rs

```diff
@@ -98,27 +104,57 @@ impl AsRef<std::net::UdpSocket> for UdpSocket {
     }
 }
 
-fn init_socket(socket: &std::net::UdpSocket) -> io::Result<()> {
-    let fd = socket.as_raw_fd();
-    let enable_ts: libc::c_int = 1;
+/// # Safety
+///
+/// The given pointer must point to a libc::timespec
+unsafe fn read_ntp_timestamp(ptr: *const u8) -> NtpTimestamp {
+    let ts: libc::timespec = std::ptr::read_unaligned(ptr as *const _);
+
+    // truncates the higher bits of the i64
+    let seconds = (ts.tv_sec as u32).wrapping_add(EPOCH_OFFSET);
+
+    // tv_nsec is always within [0, 1e10)
+    let nanos = ts.tv_nsec as u32;
+
+    NtpTimestamp::from_seconds_nanos_since_ntp_era(seconds, nanos)
+}
+
+fn set_timestamping_options(udp_socket: &std::net::UdpSocket) -> io::Result<()> {
+    let fd = udp_socket.as_raw_fd();
+
+    // our options:
+    //  - we want software timestamps to be reported,
+    //  - we want both send and receive software timestamps
+    let bits = libc::SOF_TIMESTAMPING_SOFTWARE
+        | libc::SOF_TIMESTAMPING_RX_SOFTWARE
+        | libc::SOF_TIMESTAMPING_TX_SOFTWARE;
+
     unsafe {
-        libc::setsockopt(
+        cvt(libc::setsockopt(
             fd,
             libc::SOL_SOCKET,
-            libc::SO_TIMESTAMPNS,
-            &enable_ts as *const _ as *const libc::c_void,
+            libc::SO_TIMESTAMPING,
+            &bits as *const _ as *const libc::c_void,
             std::mem::size_of::<libc::c_int>() as libc::socklen_t,
-        )
+        ))?
     };
+
     Ok(())
 }
 
+fn cvt(t: libc::c_int) -> crate::io::Result<libc::c_int> {
+    match t {
+        -1 => Err(std::io::Error::last_os_error()),
+        _ => Ok(t),
+    }
+}
+
```

Comment:
The use of `crate::io` seems really strange, not least mixed with `std::io` in the function body. Could just use `io` directly for both, right?

I'd also recommend giving this an actually meaningful name instead of `cvt`.